pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
-            if debug:
-                left, right = slits.select_edges(flexure=self.spat_flexure_shift)
-                viewer, ch = display.show_image(illum_flat, chname='illum_flat')
-                display.show_slits(viewer, ch, left, right)  # , slits.id)
-                #
-                orig_image = self.image.copy()
-                viewer, ch = display.show_image(orig_image, chname='orig_image')
-                display.show_slits(viewer, ch, left, right)  # , slits.id)
-
-        # Apply the relative spectral illumination
-        spec_illum = 1.0
-        if self.par['use_specillum']:
-            if flatimages is None or flatimages.get_spec_illum() is None:
-                msgs.error("Spectral illumination correction desired but not generated/provided.")
-            else:
-                spec_illum = flatimages.get_spec_illum().copy()
+        # TODO: Shouldn't the cosmic-ray detection/masking happen here, before
+        # flat-fielding?
 
-        # Flat field -- We cannot do illumination flat without a pixel flat (yet)
-        if self.par['use_pixelflat'] or self.par['use_illumflat']:
-            if flatimages is None or flatimages.get_pixelflat() is None:
-                msgs.error("Flat fielding desired but not generated/provided.")
-            else:
-                self.flatten(flatimages.get_pixelflat()/spec_illum, illum_flat=illum_flat, bpm=self.bpm)
+        # Calculate flexure, if slits are provided and the correction is
+        # requested.  NOTE: This step must come after trim, orient (just like
+        # bias and dark subtraction) and before field flattening.
+        self.spat_flexure_shift = None if slits is None or not self.par['spat_flexure_correct'] \
+                                    else self.spatial_flexure_shift(slits)
 
-        # Fresh BPM
-        bpm = self.spectrograph.bpm(self.filename, self.det, shape=self.image.shape)
+        if self.use_flat:
+            # Flat-field the data.  This propagates the flat-fielding
+            # corrections to the variance.
+            self.flatten(flatimages, slits=slits, debug=debug)
 
-        # Extras
-        self.build_rn2img()
-        self.build_ivar()
+        # Calculate the inverse variance
+        self.ivar = utils.inverse(self.var)
 
         # Generate a PypeItImage
         pypeitImage = pypeitimage.PypeItImage(self.image, ivar=self.ivar, rn2img=self.rn2img,
-                                              bpm=bpm, detector=self.detector,
+                                              bpm=self.bpm, detector=self.detector,
                                               spat_flexure=self.spat_flexure_shift,
-                                              PYP_SPEC=self.spectrograph.name)
+                                              PYP_SPEC=self.spectrograph.name,
+                                              units='e-' if self.par['apply_gain'] else 'ADU',
```

Comment:
I hate this ADU option. Please remove it!! Noise models only make sense in electrons. Just convert every image into counts when you read it in and then all of this ambiguity is settled. That is the natural set of units for images.

Reply:
You gave a case where you might not apply the gain in our discussion over zoom.  I've left this for now.